src/core/manager.c

```diff
@@ -1729,7 +1729,8 @@ static void invoke_sigchld_event(Manager *m, Unit *u, const siginfo_t *si) {
         unit_unwatch_pid(u, si->si_pid);
 
         if (UNIT_VTABLE(u)->sigchld_event) {
-                if (set_size(u->pids) <= 1 || iteration != u->sigchldgen) {
+                if (set_size(u->pids) <= 1 || iteration != u->sigchldgen ||
+                    u->type == UNIT_SERVICE && (si->si_pid == SERVICE(u)->main_pid || si->si_pid == SERVICE(u)->control_pid)) {
```

Comment:
Please put parentheses here to make clear what interpretation of combined `&&` and `||` you intended. Even if this expression is distinct for both you and the compiler, the next reader will scratch their head over this :)